.github/workflows/test-setup.yml

```diff
@@ -0,0 +1,246 @@
+      # race conditions.
+      group: >
+        ${{ github.repository }}-
+        ${{
+          github.event_name == 'push' && github.run_id
+          ||
+          (
+            github.event_name == 'issue_comment' && github.event.issue.number
+            ||
+            github.event.number
+          )
+        }}
+
+    steps:
+      - uses: actions/github-script@v7
+        name: Check user permissions
+        with:
+          script: |
+            // If this is a comment, immediately react with a rocket to show
+            // that things are happening
+            if (context.eventName == "issue_comment") {
+              github.rest.reactions.createForIssueComment({
+                owner: context.repo.owner,
+                repo: context.repo.repo,
+                comment_id: context.payload.comment.id,
+                content: "rocket"
+              });
+            }
+
+            // Check if the triggering actor has write access
+            let triggering_actor = ${{ toJSON(github.triggering_actor) }};
+
+            let permission_level = (await github.rest.repos.getCollaboratorPermissionLevel({
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              username: triggering_actor
+            })).data.permission;
+
+            let can_write = ["admin", "write"].includes(permission_level);
+            let is_dependabot = triggering_actor == 'dependabot[bot]';
+
+            if (!(can_write || is_dependabot)) {
+                // Post a comment and fail
+
+                // If this was a comment, mark it with a thumbs-down
+                if (context.eventName == "issue_comment") {
+                  await github.rest.reactions.createForIssueComment({
+                    owner: context.repo.owner,
+                    repo: context.repo.repo,
+                    comment_id: context.payload.comment.id,
+                    content: "-1"
+                  });
+                }
+
+                let verb = context.payload.comment.body.startsWith("/fix-tests") ? "fix" : "run";
+
+                await github.rest.issues.createComment({
+                  owner: context.repo.owner,
+                  repo: context.repo.repo,
+                  issue_number: context.payload.issue.number,
+                  body: `*Bleep bloop, I am a robot.*
+
+            Only repository collaborators are allowed to trigger the tests. Collaborators, please review the changes in this pull request and, if they are safe, ${verb} the tests on behalf of @${triggering_actor} by commenting "/${verb}-tests" (without the quotes).
+            `
+                });
+
+                core.setFailed(`${triggering_actor} does not have write permissions`);
+            }
+
+      - uses: actions/github-script@v7
+        name: Prep for test
+        with:
+          script: |
+            // Create an archive with all of the relevant control information
```

Comment:
Can github-script use external files for the scripts? We may want to consider moving them out of the YAML file for ease of maintenance and testability.

Reply:
I'm 100% with you, and in general I think _no_ - I'll look for more details, but I don't think the reusable workflow repo gets checked out.

Reply:
It seems like we could - https://github.com/orgs/community/discussions/25294 - that might be worth it.

Reply:
Also https://github.com/orgs/community/discussions/25294